initrd/directory_test.go

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2023, Unikraft GmbH and The KraftKit Authors.
+// Licensed under the BSD-3-Clause License (the "License").
+// You may not use this file except in compliance with the License.
+package initrd_test
+
+import (
+	"context"
+	"io"
+	"os"
+	"testing"
+
+	"github.com/cavaliergopher/cpio"
+
+	"kraftkit.sh/initrd"
+)
+
+func TestNewFromDirectory(t *testing.T) {
+	const rootDir = "testdata/rootfs"
+
+	ctx := context.Background()
+
+	ird, err := initrd.NewFromDirectory(ctx, rootDir)
+	if err != nil {
+		t.Fatal("NewFromDirectory:", err)
+	}
+
+	irdPath, err := ird.Build(ctx)
+	if err != nil {
+		t.Fatal("Build:", err)
+	}
+	t.Cleanup(func() {
+		if err := os.Remove(irdPath); err != nil {
+			t.Fatalf("Failed to remove initrd file %s: %v", irdPath, err)
+		}
+	})
+
+	irdFiles := ird.Files()
+
+	const expectFiles = 4 // only regular and symlink files are indexed
+	if gotFiles := len(irdFiles); gotFiles != expectFiles {
+		t.Errorf("Expected %d files in InitrdConfig, got %d: %v", expectFiles, gotFiles, irdFiles)
+	}
+
+	r := cpio.NewReader(openFile(t, irdPath))
+
+	expectHeaders := map[string]cpio.Header{
+		"/entrypoint.sh": {
```

Comment:
Not sure how the reader handles paths on Windows, I guess we can leave like this?

Reply:
sure thing

Reply:
```
[...]
go: downloading golang.org/x/text v0.12.0
--- FAIL: TestNewFromDirectory (0.02s)
    directory_test.go:83: Encountered unexpected file in cpio archive: testdata/rootfs
    directory_test.go:83: Encountered unexpected file in cpio archive: \entrypoint.sh
    directory_test.go:83: Encountered unexpected file in cpio archive: \etc
    directory_test.go:83: Encountered unexpected file in cpio archive: \etc\app.conf
    directory_test.go:83: Encountered unexpected file in cpio archive: \lib
    directory_test.go:83: Encountered unexpected file in cpio archive: \lib\libtest.so.1
    directory_test.go:83: Encountered unexpected file in cpio archive: \lib\libtest.so.1.0.0
    directory_test.go:34: Failed to remove initrd file D:\Temporary\2117131240: remove D:\Temporary\2117131240: The process cannot access the file because it is being used by another process.
FAIL
FAIL    kraftkit.sh/initrd      0.425s
FAIL
```

Reply:
This line `Failed to remove initrd file D:\Temporary\2117131240: remove D:\Temporary\2117131240: The process cannot access the file because it is being used by another process.` means that `os.Remove()` was called before all references to the file were closed

Reply:
Probably this happens because the first cleanup registered is the one that removes the file, and the second one the one that closes it, so the order is reversed

Reply:
Thanks for the feedback!

The fact that CPIO archives contain Windows paths is a bug, not a feature ;)

Also as you can see, the constant value doesn't need to change. Go still found the testdata.

Reply:
Considering that they are used inside unikraft, which has Linux paths, you are right

Should you do this, or should @nderjung fix and I approve afterwards?

Also don't forget about the reverse order thingy

Reply:
```
PS C:\Users\Cezar\kraftkit> gh pr checkout 953 --force
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 4 (delta 2), reused 4 (delta 2), pack-reused 0
Unpacking objects: 100% (4/4), 1.63 KiB | 104.00 KiB/s, done.
From https://github.com/unikraft/kraftkit
 * branch            refs/pull/953/head -> FETCH_HEAD
HEAD is now at 0edcb4d test(initrd): assert contents of CPIO archive
PS C:\Users\Cezar\kraftkit> go test ./initrd
--- FAIL: TestNewFromDirectory (0.01s)
    directory_test.go:83: Encountered unexpected file in cpio archive: testdata/rootfs
    directory_test.go:83: Encountered unexpected file in cpio archive: \entrypoint.sh
    directory_test.go:83: Encountered unexpected file in cpio archive: \etc
    directory_test.go:83: Encountered unexpected file in cpio archive: \etc\app.conf
    directory_test.go:83: Encountered unexpected file in cpio archive: \lib
    directory_test.go:83: Encountered unexpected file in cpio archive: \lib\libtest.so.1
    directory_test.go:83: Encountered unexpected file in cpio archive: \lib\libtest.so.1.0.0
    directory_test.go:34: Failed to remove initrd file: remove D:\Temporary\1822746801: The process cannot access the file because it is being used by another process.
FAIL
FAIL    kraftkit.sh/initrd      0.407s
FAIL
```

Reply:
are you sure these are defers though, it uses `t.Cleanup` which maybe has different functionality?

Reply:
Cleanup works the same as defer. It would fail in the CI (on Linux) if the code was problematic, but right now it fails only on Windows.

Reply:
Those errors are not checked in Kraftkit:

https://github.com/unikraft/kraftkit/blob/0d09e3964273b7e5d9e7e60f254782b15ff076ca/initrd/directory.go#L68-L71

For all that I know, the problem could be there.

Reply:
definitely possible 🤷